### PR TITLE
Deprecate apply_theta_transforms=True to PolarTransform

### PR DESCRIPTION
**Pull request by dstansby**
## PR Summary
This parameter was introduced in https://github.com/matplotlib/matplotlib/pull/4699 for backwards compatabilty reasons, when shifting theta values was refactored out of `PolarTransform`.

- It think it's reasonable to **not** expect `PolarTransform` to do a theta shift first
- We do not currently test the case `_apply_theta_transforms = True`
- Removing this will simplify the polar transform code slightly

So I think it's worth deprecating and eventually removing this.

## PR Checklist

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by anntzer** [comment]
InvertedPolarTransform also needs this?

**Comment by tacaswell** [comment]
Pushing to 3.8 as I do not think this is critical to get in and is still draft.  Please push back if I am incorrect @dstansby .

**Comment by tacaswell** [comment]
We talked about this on the call this week and we felt we did not have enough context to understand why we want to do this and concern about exposing an underscored arg name to the users.

There was also surprise about the claim that this was not tested given it is the default!

Do we want to make this public if we are going to document it like this?  Should we just have a second class?

**Comment by jklymak** [comment]
I guess the point is that a polar axes sets this to False, which is our internal use.  So this would only affect power users who are using PolarTransform directly?  OTOH, it seems we should have a public toggle for this, and document it?

**Comment by dstansby** [comment]
I think the original intention was to keep the 'version' of the class where no theta transform was applied private, but for backwards API compatability keep a version where a theta transform was applied. So I think making the argument public and going through a deprecation as is in this PR might be a reasonable way forward?

The motivation for this was to make `PolarTransform` internal code less complicated for maintenance reasons. I figured that there probably aren't lots of people out there using this, and there's a clear and simple alternative to `_apply_theta_transforms=True` (just add a theata value to the outputs), so it was worth doing.

**Review by jklymak — APPROVED** [review]
I think this is OK since this is probably only used by developers, if anyone.  Maybe not the nicest dance, but improves maintainability.

File: doc/api/next_api_changes/deprecations/24834-DS.rst [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+but only takes affect when:
+
+- An axis is associated with the transform.
+- The axis has a non-zero theta offset or has theta values increasing in
+  a clockwise direction.
+
+To silence this warning and adopt future behaviour,
+set ``apply_theta_transforms=False``. If you need to retain the behaviour
+where theta values are transformed, chain the ``PolarTransform`` with
+another transform that performs the theta shift and/or sign shift.
```
Comment (QuLogic) [inline doc/api/next_api_changes/deprecations/24834-DS.rst:16]: Maybe add an example of which transform to use?

File: lib/matplotlib/projections/polar.py [hunk 2]
```diff
@@ -15,6 +15,19 @@
 from matplotlib.spines import Spine
 
 
+def _apply_theta_transforms_warn():
+    _api.warn_deprecated(
+                "3.7",
```
Comment (QuLogic) [inline lib/matplotlib/projections/polar.py:20]:
```suggestion
                "3.8",
```

File: lib/matplotlib/projections/polar.py [hunk 3]
```diff
@@ -15,6 +15,19 @@
...
+                message=(
+                    "Passing `apply_theta_transforms=True` (the default) "
+                    "is deprecated. Support for this will be removed in "
+                    "Matplotlib %(removal)s. To prevent this warning, "
+                    "set `apply_theta_transforms=False`, and make sure to "
+                    "shift theta values before being passed to this transform."
+                )
```
Comment (QuLogic) [inline lib/matplotlib/projections/polar.py:27]: There should be a `%(since)s` in the message as well.

**Review by QuLogic — COMMENTED** [review]
This needs a rebase now.

File: lib/matplotlib/projections/polar.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/projections/polar.py:20]: This should be 3.9 now.

Comment (dstansby) [reply lib/matplotlib/projections/polar.py]: 👍 updated

File: lib/matplotlib/projections/polar.py (hunk 3 above)
Comment (dstansby) [reply lib/matplotlib/projections/polar.py]: 👍  added

File: doc/api/next_api_changes/deprecations/24834-DS.rst (hunk 1 above)
Comment (dstansby) [reply doc/api/next_api_changes/deprecations/24834-DS.rst]: 👍 added a note that an `Affine2D` can be used to do this.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**